### PR TITLE
[RFC] [DO NOT MERGE] Ensure link DMA channel with the same ID as trace host DMA is reserved

### PR DESCRIPTION
**Pull request by ranj063**
There is a HW limitation that requires the host and link DMA engines
to be coupled whenever link DMA engine needs to be reset. This means
that even in decoupled mode, we always need to reserve link and
host DMA channels with the same ID. Trace utilizes only the host DMA.
Therefore, to prevent the link DMA with the same ID as trace host DMA
from being assigned to another stream, its link_locked attribute
must be set.

fixes thesofproject/sof#1173

### CODE REVIEW

**Comment by ranj063** [comment]
@RanderWang @keqiaozhang I dont have a CML to test this. Could I please request to check if this fixes the issue?

**Comment by RanderWang** [comment]
> @RanderWang @keqiaozhang I dont have a CML to test this. Could I please request to check if this fixes the issue?

This PR could fix the failure of FW loading, but it doesn't fix the coupled & decoupled issue in SOF.

**Comment by RanderWang** [comment]
I share you some cases to your list. 
There are two issue in our SOF. one is the coupled-set_format-decoupled sequence in hda_stream_params, which would coupled alink dma channel used by other streams. Another one is restoring coupled mode when stream is closed, and this would also coupled a link dma channel used by other streams. I assume you get these points. It is also discussed in our Email.

 Assuming all dma channel zero are used by dma trace.
(1) pcm1 is running with host dma channel two and link dma channel one, and pcm2 is started with host dma channel one and link dma channel two. (this is a bug report by QA) The coupled-set_format-decoupled sequence in hda_stream_params for pcm2 would break pcm1 because it would couple host dma channel one and link dma channel one used by pcm1.  And even without this issue, when pcm2 is closed, it would restore link dma one & host dma one to coupled mode, this would also break pcm1

(2) a pcm with one FE and multiple BEs

(3) There are one FE and two BEs for pcm1,  one FE & one BE fro pcm2, how to make them not affect each other. The key idea here is how to allocated dma channel

(3) a few pcms  mixed by FW to one BEs.

(4)  for tone generator, there is no FE only BE

(5) for stream like dma trace, there is only FE without BE

(6) For M streams map to N BEs. The streams are enabled in different time and maybe disabled in different time.

My advice is to refine the allocation algorithm and refine coupled mode setting in hw_free. According to Proborszcz, Filip( please review the Email for coupled &decoupled mode), they also refined the allocation algorithm on windows.

**Comment by ranj063** [comment]
@RanderWang ack and thanks for the explanation. Agree that we need a slightly more sophisticated stream allocation scheme to get around the HW limitation. But this PR is still relevant isnt it?

**Comment by RanderWang** [comment]
> @RanderWang ack and thanks for the explanation. Agree that we need a slightly more sophisticated stream allocation scheme to get around the HW limitation. But this PR is still relevant isnt it?

yes, it is relevant. It is a special case for allocation algorithm.

File: sound/soc/sof/intel/hda-trace.c [hunk 1]
```diff
@@ -51,6 +51,17 @@ int hda_dsp_trace_init(struct snd_sof_dev *sdev, u32 *stream_tag)
 		return -ENODEV;
 	}
 
+	/*
+	 * There is a HW limitation that requires the host and link DMA engines
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-trace.c:55]: nope. the hw recommended programming sequence is to couple the host and link DMA so set the format, as done in a previous commit. This change has nothing to do with reset.

**Comment by ranj063** [comment]
@plbossart I understand this isnt generic enough. I think this change along with reverting back to assigning the link dma channel during hw_params should take care of most of the cases. The only case that will still not work is 1 FE connected to multiple BE's. I think I'll need to have a chat with you about this.

**Comment by ranj063** [comment]
Closing this one. Will be replaced by a slightly more generic solution soon

**Pull request closed without merging**